File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionNumberParse.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Locale;
import java.util.Map;

/*
 * For tests
 * @see TransformerFunctionNumberParseTest
 */
public class TransformerFunctionNumberParse extends TransformerFunction {
    public TransformerFunctionNumberParse() {
        super(FunctionDescription.of(
                Map.of(
                        "pattern", ArgumentType.of(ArgType.String).position(0).defaultString("#0.00"),
                        "locale", ArgumentType.of(ArgType.String).position(1).defaultIsNull(true),
                        "grouping", ArgumentType.of(ArgType.String).position(2).defaultIsNull(true),
                        "decimal", ArgumentType.of(ArgType.String).position(3).defaultIsNull(true),
                        "radix", ArgumentType.of(ArgType.Integer).position(1).defaultInteger(10)
                )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        var pattern = context.getString("pattern");
        if (pattern == null) {
            return new BigDecimal(str);
        }
        if ("BASE".equals(pattern)) {
            var radix = context.getInteger("radix");
            return new BigDecimal(new BigInteger(str, radix));
        }

        var locale = context.getString("locale");
        var resolvedLocale = FunctionHelpers.isNullOrEmpty(locale)
                             ? FunctionHelpers.DEFAULT_LOCALE
                             : Locale.forLanguageTag(locale);
        var grouping = context.getString("grouping");
        var decimal = context.getString("decimal");

        var formatter = FunctionHelpers.getDecimalFormatter(resolvedLocale, pattern, grouping, decimal);
        try {
            return formatter.parse(str);
        } catch (Throwable t) {
            throw t instanceof RuntimeException ex ? ex : new RuntimeException(t);
        }
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionNumberParse.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { BigDecimal, DEFAULT_LOCALE } from "./common/FunctionHelpers";
import { isNullOrEmpty } from "../JsonHelpers";
import { parsePattern } from "./TransformerFunctionNumberFormat";

class TransformerFunctionNumberParse extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        pattern: { type: ArgType.String, position: 0, defaultString: "#0.00" },
        locale: { type: ArgType.String, position: 1, defaultIsNull: true },
        grouping: { type: ArgType.String, position: 2, defaultIsNull: true },
        decimal: { type: ArgType.String, position: 3, defaultIsNull: true },
        radix: { type: ArgType.Integer, position: 1, defaultInteger: 10 },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    const pattern = await context.getString("pattern");
    if ("BASE" === pattern) {
      const radix = await context.getInteger("radix");
      return BigDecimal(str, radix ?? 10);
    }

    const locale = await context.getString("locale");
    const resolvedLocale = isNullOrEmpty(locale) ? DEFAULT_LOCALE : locale;
    const groupSeparator = Intl.NumberFormat(resolvedLocale)
      .format(11111)
      .replace(/\p{Number}/gu, "");
    const decimalSeparator = Intl.NumberFormat(resolvedLocale)
      .format(1.1)
      .replace(/\p{Number}/gu, "");

    let normalized = str;
    if (pattern != null) {
      const [format] = parsePattern(pattern, true /* usually negative is the special one */);
      if (format.prefix && normalized.startsWith(format.prefix)) {
        normalized = normalized.substring(format.prefix.length);
      }
      if (format.suffix && normalized.endsWith(format.suffix)) {
        normalized = normalized.substring(0, normalized.length - format.suffix.length);
      }
    }

    const grouping = (await context.getString("grouping")) ?? groupSeparator;
    const decimal = (await context.getString("decimal")) ?? decimalSeparator;

    normalized = normalized.replace(new RegExp(`[${grouping}]`, "g"), "");
    if (decimal !== "." && decimal) {
      normalized = normalized.replace(decimal, ".");
    }

    return BigDecimal(normalized);
  }
}

export default TransformerFunctionNumberParse;
